File: jan-22-25/sub.cpp
#include <vector>
using namespace std;

int q[1000000];

class Solution {
public:
    vector<vector<int>> highestPeak(vector<vector<int>>& isWater) {
        int rows = isWater.size(), cols = isWater[0].size();
        int fetch = 0, put = 0;
        int drow[] = {-1,1,0,0};
        int dcol[] = {0,0,-1,1};

        for(int i = 0; i<rows; i++){
            for(int j = 0; j<cols; j++){
                if(isWater[i][j] == 1) q[put++] = i * cols + j;
                isWater[i][j]--;
            }
        }

        int pos, row, col, val, nrow, ncol;

        while(put > fetch){
            pos = q[fetch++];
            row = pos / cols;
            col = pos % cols;
            val = isWater[row][col];

            for(int i = 0; i<4; i++){
                nrow = row + drow[i];
                ncol = col + dcol[i];

                if(nrow >= 0 && nrow < rows && ncol >= 0 && ncol < cols && isWater[nrow][ncol] == -1){
                    isWater[nrow][ncol] = val + 1;
                    q[put++] = nrow * cols + ncol;
                }
            }
        }

        return isWater;
    }
};

File: jan-22-25/sub.py
from typing import List
from collections import deque

class Solution:
    def highestPeak(self, isWater: List[List[int]]) -> List[List[int]]:
        
        rows = len(isWater)
        cols = len(isWater[0])

        q = deque()

        dirs = [(-1,0), (1,0), (0,-1), (0,1)]

        for i in range(rows):
            for j in range(cols):
                if isWater[i][j] == 1:
                    isWater[i][j] = 0
                    q.append(i * cols + j)
                else:
                    isWater[i][j] = -1


        while q:
            
            pos = q.popleft()
            row = pos//cols
            col = pos % cols
            val = isWater[row][col]

            for drow, dcol in dirs:

                nrow, ncol = row + drow, col + dcol

                if 0 <= nrow < rows and 0 <= ncol < cols and isWater[nrow][ncol] == -1:

                    isWater[nrow][ncol] = val + 1

                    q.append(nrow * cols + ncol)

        return isWater

